checker/src/main/java/org/checkerframework/checker/mustcall/MustCallAnnotatedTypeFactory.java

```diff
@@ -0,0 +1,363 @@
+    } else {
+      throw new BugInCF("unexpected type of method tree: " + tree.getKind());
+    }
+    changeNonOwningParameterTypesToTop(declaration, type);
+    super.methodFromUsePreSubstitution(tree, type);
+  }
+
+  @Override
+  protected void constructorFromUsePreSubstitution(
+      NewClassTree tree, AnnotatedExecutableType type) {
+    ExecutableElement declaration = TreeUtils.elementFromUse(tree);
+    changeNonOwningParameterTypesToTop(declaration, type);
+    super.constructorFromUsePreSubstitution(tree, type);
+  }
+
+  /**
+   * Changes the type of each parameter not annotated as @Owning to top. Also replaces the component
+   * type of the varargs array, if applicable.
+   *
+   * <p>Note that this method is not responsible for handling receivers, which can never be owning.
+   *
+   * @param declaration a method or constructor declaration
+   * @param type the method or constructor's type
+   */
+  private void changeNonOwningParameterTypesToTop(
+      ExecutableElement declaration, AnnotatedExecutableType type) {
+    for (int i = 0; i < type.getParameterTypes().size(); i++) {
+      Element paramDecl = declaration.getParameters().get(i);
+      if (checker.hasOption(MustCallChecker.NO_LIGHTWEIGHT_OWNERSHIP)
+          || getDeclAnnotation(paramDecl, Owning.class) == null) {
+        AnnotatedTypeMirror paramType = type.getParameterTypes().get(i);
+        if (!paramType.hasAnnotation(POLY)) {
+          paramType.replaceAnnotation(TOP);
+        }
+        if (declaration.isVarArgs() && i == type.getParameterTypes().size() - 1) {
+          // also modify the component type of a varargs array
+          AnnotatedTypeMirror varargsType = ((AnnotatedArrayType) paramType).getComponentType();
+          if (!varargsType.hasAnnotation(POLY)) {
+            varargsType.replaceAnnotation(TOP);
+          }
+        }
+      }
+    }
+  }
+
+  @Override
+  public AnnotatedTypeMirror fromElement(Element elt) {
+    AnnotatedTypeMirror type = super.fromElement(elt);
+    // Support @InheritableMustCall meaning @MustCall on all class declaration elements.
+    if (ElementUtils.isTypeElement(elt)) {
+      AnnotationMirror inheritableMustCall = getDeclAnnotation(elt, InheritableMustCall.class);
+      if (inheritableMustCall != null) {
+        List<String> mustCallVal =
+            AnnotationUtils.getElementValueArray(
+                inheritableMustCall, inheritableMustCallValueElement, String.class);
+        AnnotationMirror inheritedMCAnno = createMustCall(mustCallVal.toArray(new String[0]));
+        // Ensure that there isn't an inconsistent, user-written @MustCall annotation and
+        // issue an error if there is. Otherwise, replace the implicit @MustCall({}) with
+        // the inherited must-call annotation.
+        AnnotationMirror writtenMCAnno = type.getAnnotationInHierarchy(this.TOP);
+        if (writtenMCAnno != null
+            && !this.getQualifierHierarchy().isSubtype(inheritedMCAnno, writtenMCAnno)) {
```

Comment:
Why is this a subtype test rather than an equality test?  Why is it OK to have an `@InheritableMustCall` and a `@MustCall` annotation with different values?

Reply:
For that same reason that subtyping between must-call annotations is generally permitted. A programmer can always write a more-restrictive `@MustCall` annotation on a class than the `@MustCall` annotation on that class' superclass; otherwise, you wouldn't be able to e.g. write `@MustCall("close")` on `java.net.Socket`, because it extends `Object`, which is implicitly `@MustCall({})`.